### ISSUE TITLE
Mistaken type checking error

### ISSUE THREAD

**Issue by carlocab** [rhysd/actionlint#294]
I'm getting the following error which I believe to be mistaken:
```
object, array, and null values should not be evaluated in template with ${{ }} but evaluating the value of type array<any> [expression]
```
[Here](https://rhysd.github.io/actionlint/#eJxNjsEOgjAQRO98xR5MSEzoB/TqyRvhBwg1K6ClxXZXNIR/t7QQOb6Z3ZmxRmYAk3XPu7ZT7TjyX/EJAQq4XLOHVVEwPCh0mxd+fGGNBFZsiAvdEHqKlicXoP3uIUNDrv/sBFBWEk7zDG1PHSuBbzQkjmPEyFrXDl8cIr04i9QMy7Ll43hYuK4HvHUW8jU1tYmyCuf5D5T9SFM=)'s a minimal reproducer.

The value of `matrix.PR` is, of course, an array. actionlint doesn't seem to recognise here that I will have a different job dispatched for each entry in the `matrix.PR` array, and each job will only see one value for `matrix.PR`.

**Comment by rhysd** [comment]
The error looks correct. The `PR` matrix value is an array because it is generated by filter syntax (`github.event.workflow_run.pull_requests.*.number`). An array value can not be embedded with `${{ }}` syntax directly. I think what you want to do is:

```yaml
- run: echo '${{ toJSON(matrix.PR) }}'
```

**Comment by carlocab** [comment]
But `matrix.PR` is a scalar when evaluated inside the `run:` statement because of the way GitHub Actions handles `matrix`, so doing
```yaml
- run: echo '${{ matrix.PR }}'
```
is not an error.

**Comment by rhysd** [comment]
> But matrix.PR is a scalar when evaluated inside the run: statement

That's because the result of the filter syntax happened to be an array with a single element, am I correct? If the result contained two elements, the result would've been different.

**Comment by carlocab** [comment]
> That's because the result of the filter syntax happened to be an array with a single element, am I correct? 

No, that is not correct.

> If the result contained two elements, the result would've been different.

`github.event.workflow_run.pull_requests.*.number` can contain any number of elements. Regardless of the number of elements this array contains, in the context of `run:`, `matrix.PR` is always a scalar (conditional on the `run:` statement being reached).

Please see: https://docs.github.com/en/actions/using-jobs/using-a-matrix-for-your-jobs

**Comment by carlocab** [comment]
Thanks, @rhysd!
